File: nim/src/coding_practice/chapter1_4.nim
# Write a method to decide if two strings are anagrams or not.

import math

func are_anagrams*(str1, str2: string): bool =
  var seen: array[128, int]
  for ch in str1:
    seen[int(ch)] += 1
  for ch in str2:
    seen[int(ch)] -= 1
    if seen[int(ch)] < 0: return false
  if sum(seen) == 0: return true
  false

File: elixir/lib/chapter1_4.ex
defmodule Crack.Chapter1_4 do
  @moduledoc """
  1.4 Anagrams:

  Write a method to decide if two strings are anagrams or not.
  """

  @doc """
  Checks if two strings are anagrams.

      iex> Crack.Chapter1_4.start("hello", "olelh")
      true

      iex> Crack.Chapter1_4.start("arg", "?")
      false

      iex> Crack.Chapter1_4.start("jakob beckmann", "manja bbocenkk")
      true

      iex> Crack.Chapter1_4.start("hello", "")
      false

      iex> Crack.Chapter1_4.start("holla", "hello")
      false

  """
  def start(str1, str2) when is_binary(str1) and is_binary(str2) do
    chars_1 = str1 |> String.graphemes |> Enum.sort
    chars_2 = str2 |> String.graphemes |> Enum.sort
    chars_1 === chars_2
  end
end
